_posts/2019-02-26-pozycjonowanie-zale-ne-od-scrolla.md

```diff
@@ -0,0 +1,102 @@
+---
+layout:    post
+title:     "Pozycjonowanie zależne od scrolla"
+date:      2019-02-26 08:00:00 +0100
+published: true
+author:    glipecki
+tags:
+    - tech
+    - frontend
+    - javascript
+---
+
+Czasem zachodzi potrzeba uzależnienia pozycji elementu od scrolla okna, czy to na potrzeby przyklejenia w widocznym obszarze, czy też stworzenia efektu paralaksy lub niestandardowego flow nawigacji. Temat wydawałby się oczywisty, gdyby nie to, że celowo wprowadzimy sobie dodatkowe ograniczenia (co wcale nie jest takie niecodziennie, uwzględniając fantazję działów UI/UX ;-)).
+
+## Podejście 1: tylko CSS
+Mamy dwa sposoby przyklejenia elementu do ekranu wykorzystujące tylko CSS. Oba opierają się o zmianę pozycjonowania:
+
+* `position: fixed`,
+* `position: sticky`.
+
+Oba też mają swoje problemy i ograniczenia.
+
+Stosując pozycjonowanie fixed:
+* musimy uwzględnić pozostawione przez niego miejsce w oryginalnym fragmencie drzewa DOM,
+* jeżeli element będzie wyższy niż viewport to nie będziemy mieli możliwości obejrzeć niemieszczącej się zawartości,
+* fixed zawsze tworzy nowy stacking context.
+
+Stosując pozycjonowanie sticky:
+* jeżeli element będzie wyższy niż viewport, to nie będziemy mieli możliwości obejrzeć niemieszczącej się zawartości,
+* sticky zawsze tworzy nowy stacking context,
+* sticky czasem może zaskoczyć swoim działaniem (przykładowo [issue w3c](https://github.com/w3c/csswg-drafts/issues/865)).
+
+O ile uwzględnienie oderwanego przez fixed elementu w layoucie nie stanowi wyzwania, o tyle brak wsparcia dla przewijania treści i zmiana stacking context (co wpłynie np. na liczenie kolejności na osi z) mogą stanowić już zbyt duże ograniczenia.
+
+W wielu przypadkach fixed lub sticky załatwią problem. Jeśli jednak potrzebujesz czegoś więcej, czytaj dalej.
+
+## Podejście 2: JavaScript
+„Nie ma takiej rzeczy, której bym nie napisał w JavaScript.” 😉
+
+Przeglądarki oferują nam zdarzenie związane ze scrollowaniem treści. Na zdarzenie możemy nasłuchiwać przez zdefiniowanie własności `target.onscroll`, czy też bardziej elastycznie, dodając listener przez `target.addEventListner(’scroll’)`. Teoretycznie wystarczyłoby już tylko przeliczać pozycję przyklejanego elementu, obsłużyć przewijanie w dwóch kierunkach i nie zapomnieć o użyciu najmniej obciążającej metody przesuwania elementów po ekranie. Co może pójść źle? Sprawdźmy prosty przykład.
+
+Implementujemy proste przeliczanie pozycji nasłuchując na zdarzenie scroll:
+
+<iframe width="100%" height="300" src="//jsfiddle.net/gregorry/gof6we57/embedded/" allowfullscreen="allowfullscreen" allowpaymentrequest frameborder="0"></iframe>
+
+Uzyskany efekt:
+
+<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/86dZ7HguQWY" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

Comment:
To będzie słabo wyglądać na mobilach. Wrzuciłem template dla iframe youtube'a, można go użyć w ten sposób:
{% include youtube.html movie="86dZ7HguQWY" %}

Reply:
super, dzięki!

Reply:
a iframe z kodami też krzaczy?

Reply:
done

Reply:
Kody wyglądają spoko, co prawda scroll się pojawia w trochę słabym miejscu, ale z tym za dużo nie zrobimy